examples/platform/openiotsdk/app/openiotsdk_startup_gcc.cpp

```diff
@@ -0,0 +1,240 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <errno.h>
+#include <new>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "cmsis_os2.h"
+#include "hal/serial_api.h"
+
+#ifdef TFM_SUPPORT
+extern "C" uint32_t tfm_ns_interface_init(void);
+#endif // TFM_SUPPORT
+
+#define CALLER_ADDR() __builtin_extract_return_addr(__builtin_return_address(0))
+
+// Consider reducing the baudrate if the serial is used as input and characters are lost
+extern "C" mdh_serial_t * get_example_serial();
+#ifndef IOT_SDK_APP_SERIAL_BAUDRATE
+#define IOT_SDK_APP_SERIAL_BAUDRATE 921600
+#endif
+
+// main thread declaration
+// The thread object and associated stack are statically allocated
+#ifndef IOT_SDK_APP_MAIN_STACK_SIZE
+#define IOT_SDK_APP_MAIN_STACK_SIZE 16 * 1024
+#endif
+static void main_thread(void * argument);
+alignas(8) char main_thread_stack[IOT_SDK_APP_MAIN_STACK_SIZE];
+alignas(8) uint8_t main_thread_storage[100] __attribute__((section(".bss.os.thread.cb")));
+
+// malloc mutex declaration
+osMutexId_t malloc_mutex;
+alignas(8) uint8_t malloc_mutex_obj[80];
+
+// C runtime import: constructor initialization and main
+extern "C" void __libc_init_array(void);
+extern "C" int main(void);
+
+/*
+ * This function override startup sequence. Instead of releasing control to the C runtime
+ * the following operations are performed:
+ *
+ * - initialize the serial (low level)
+ * - initialize RTOS
+ * - Start the RTOS with the main thread
+ */
+extern "C" void mbed_sdk_init(void)
+{
+    mdh_serial_set_baud(get_example_serial(), IOT_SDK_APP_SERIAL_BAUDRATE);
+
+    int ret = osKernelInitialize();
+    if (ret != osOK)
+    {
+        printf("osKernelInitialize failed: %d\r\n", ret);
+        abort();
+    }
+
+    // Create main thread used to run the application
+    {
+        osThreadAttr_t main_thread_attr = {
+            .name       = "main",
+            .cb_mem     = &main_thread_storage,
+            .cb_size    = sizeof(main_thread_storage),
+            .stack_mem  = main_thread_stack,
+            .stack_size = sizeof(main_thread_stack),
+            .priority   = osPriorityNormal,
+        };
+
+        osThreadId_t main_thread_id = osThreadNew(main_thread, NULL, &main_thread_attr);
+        if (main_thread_id == NULL)
+        {
+            printf("Main thread creation failed\r\n");
+            abort();
+        }
+    }
+
+    ret = osKernelStart();
+    // Note osKernelStart should never return
+    printf("Kernel failed to start: %d\r\n", ret);
+    abort();
+}
+
+/**
+ * Main thread
+ * - Initialize TF-M
+ * - Initialize the toolchain:
+ *  - Setup mutexes for malloc and environment
+ *  - Construct global objects
+ * - Run the main
+ */
+static void main_thread(void * argument)
+{
+#ifdef TFM_SUPPORT
+    {
+        int ret = tfm_ns_interface_init();
```

Comment:
Why do we need TF-M initialization in this place?

Reply:
It make sense to me to have initialisation of low level component before constructors are initialised. 
I'd prefer to avoid the situation where something in chip crypto requires PSA and calls it in a constructor before the system has been properly initialised.